File: hrms-backend/src/main/java/com/hrms/domain/recognition/Recognition.java
package com.hrms.domain.recognition;

import com.hrms.common.entity.TenantAware;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "recognitions")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Recognition extends TenantAware {


    @Column(nullable = false)
    private UUID giverId;

    @Column(nullable = false)
    private UUID receiverId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RecognitionType type;

    @Enumerated(EnumType.STRING)
    private RecognitionCategory category;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String message;

    @Builder.Default
    private Integer pointsAwarded = 0;

    @Builder.Default
    private Boolean isPublic = true;

    @Builder.Default
    private Boolean isAnonymous = false;

    private UUID badgeId;

    @Builder.Default
    private Integer likesCount = 0;

    @Builder.Default
    private Integer commentsCount = 0;

    @Builder.Default
    private Boolean isApproved = true;

    private UUID approvedBy;
    private LocalDateTime approvedAt;

    private LocalDateTime recognizedAt;

    public enum RecognitionType {
        KUDOS,
        APPRECIATION,
        ACHIEVEMENT,
        MILESTONE,
        SPOT_AWARD,
        PEER_NOMINATION,
        MANAGER_RECOGNITION,
        TEAM_RECOGNITION
    }

    public enum RecognitionCategory {
        TEAMWORK,
        INNOVATION,
        CUSTOMER_FOCUS,
        LEADERSHIP,
        PROBLEM_SOLVING,
        GOING_EXTRA_MILE,
        MENTORSHIP,
        QUALITY,
        INITIATIVE,
        COLLABORATION,
        COMMUNICATION,
        OTHER
    }

    public void approve(UUID approverId) {
        this.isApproved = true;
        this.approvedBy = approverId;
        this.approvedAt = LocalDateTime.now();
    }

    public void incrementLikes() {
        this.likesCount++;
    }

    public void decrementLikes() {
        if (this.likesCount > 0) {
            this.likesCount--;
        }
    }

    public void incrementComments() {
        this.commentsCount++;
    }
}

File: frontend/lib/types/recognition.ts
// Recognition Types

export enum RecognitionType {
  KUDOS = 'KUDOS',
  APPRECIATION = 'APPRECIATION',
  ACHIEVEMENT = 'ACHIEVEMENT',
  MILESTONE = 'MILESTONE',
  SPOT_AWARD = 'SPOT_AWARD',
  PEER_NOMINATION = 'PEER_NOMINATION',
  MANAGER_RECOGNITION = 'MANAGER_RECOGNITION',
  TEAM_RECOGNITION = 'TEAM_RECOGNITION',
}

export enum RecognitionCategory {
  TEAMWORK = 'TEAMWORK',
  INNOVATION = 'INNOVATION',
  CUSTOMER_FOCUS = 'CUSTOMER_FOCUS',
  LEADERSHIP = 'LEADERSHIP',
  PROBLEM_SOLVING = 'PROBLEM_SOLVING',
  GOING_EXTRA_MILE = 'GOING_EXTRA_MILE',
  MENTORSHIP = 'MENTORSHIP',
  QUALITY = 'QUALITY',
  INITIATIVE = 'INITIATIVE',
  COLLABORATION = 'COLLABORATION',
  COMMUNICATION = 'COMMUNICATION',
  OTHER = 'OTHER',
}

export enum ReactionType {
  LIKE = 'LIKE',
  LOVE = 'LOVE',
  CELEBRATE = 'CELEBRATE',
  SUPPORT = 'SUPPORT',
  INSIGHTFUL = 'INSIGHTFUL',
}

export interface Recognition {
  id: string;
  giverId: string;
  giverName?: string;
  receiverId: string;
  receiverName?: string;
  type: RecognitionType;
  category?: RecognitionCategory;
  title: string;
  message?: string;
  pointsAwarded: number;
  isPublic: boolean;
  isAnonymous: boolean;
  badgeId?: string;
  badgeName?: string;
  likesCount: number;
  commentsCount: number;
  isApproved: boolean;
  approvedBy?: string;
  approvedAt?: string;
  recognizedAt: string;
  createdAt: string;
}

export interface RecognitionRequest {
  receiverId: string;
  type: RecognitionType;
  category?: RecognitionCategory;
  title: string;
  message?: string;
  pointsAwarded?: number;
  isPublic?: boolean;
  isAnonymous?: boolean;
  badgeId?: string;
}

export interface RecognitionBadge {
  id: string;
  name: string;
  description?: string;
  iconUrl?: string;
  pointsValue: number;
  isActive: boolean;
  category?: string;
}

export interface EmployeePoints {
  id: string;
  employeeId: string;
  employeeName?: string;
  totalPointsEarned: number;
  currentBalance: number;
  pointsGiven: number;
  recognitionsReceived: number;
  recognitionsGiven: number;
  rank?: number;
}

export interface Milestone {
  id: string;
  employeeId: string;
  employeeName?: string;
  milestoneType: string;
  title: string;
  description?: string;
  achievedDate?: string;
  upcomingDate?: string;
}

export interface EngagementDashboard {
  totalRecognitions: number;
  recognitionsThisMonth: number;
  totalPointsAwarded: number;
  activeUsers: number;
  topReceivers: EmployeePoints[];
  topGivers: EmployeePoints[];
  recognitionsByCategory: Record<string, number>;
  recentRecognitions: Recognition[];
}
